### ISSUE TITLE
More on SHA-1s

### ISSUE THREAD

**Issue by justgo129** [USGCRP/gcis-ontology#19]
Given today's discussion, I've noticed that foaf:sha1 is a property.  However, the description implies that this term is "unstable" and "not coherent."  @zednis , please clarify what this means and advise whether that would be an acceptable replacement for dbpedia:SHA-1 in our context of computer files.

See:
http://xmlns.com/foaf/spec/#term_sha1
https://data.globalchange.gov/file/9bdd68ab-4b90-4ad9-b043-fedb7e564184.thtml

Thanks.

**Comment by zednis** [comment]
I would not use `foaf:sha1` because of its instability.

`dbpedia:SHA-1` is an instance, not a property.  we could use it as the object of a property used to designate which hash function was used.

What we want is a property for cryptographic hash or similar.  If we can't find one (looking around now) we could define our own.

here is an example:

```
<http://data.globalchange.gov/file/9bdd68ab-4b90-4ad9-b043-fedb7e564184>
   dcterms:identifier "9bdd68ab-4b90-4ad9-b043-fedb7e564184";
   gcis:cryptographicHash [ 
      gcis:hashFunction dbpedia:SHA-1;
      rdf:value "bdbe9f11d4b77589c541c062cfdeca636440dd82" ] ;
   dcterms:format <http://purl.org/NET/mediatypes/application/pdf> ;
   schema:isRelatedTo <http://data.globalchange.gov/report/2014car>;
```

`dbpedia:MIME` is also a resource/instance and not a property.  I have replaced it with `dcterms:format` which is an object property.  Examples of proper usage of `dcterms:format` are available at http://wiki.dublincore.org/index.php/User_Guide/Publishing_Metadata#dcterms:format

**Comment by aulenbac** [comment]
Just curious: do we want and/or need a cryptographic hash in the GCIS ontology? Do we have a sense of how and when this would be used as part of an ontology? @abuddenb and @bduggan, what are your thoughts?

**Comment by abuddenb** [comment]
In the context of input and output files?  It would ensure the highest level of traceability but I worry about the practicality of collecting it--same issues as assigning identifiers to files.  Also, not everyone uses SHA1.

**Comment by zednis** [comment]
Cryptographic hashes (any function) can be collected and included in the RDF on a best-effort basis.

With the modeling above we can also support other hash function types, the templates will just have to have enough information to disambiguate different hash function types.

**Comment by aulenbac** [comment]
Agreed it can be collected on a best-effort basis but do we want and/or need it? How and when would a user use a cryptographic hash as part of an ontology? @abuddenb makes several good points. Do we need to support other hash types in either the relational or the semantic model? I am certainly fine if we feel this is needed. I am thinking about what we need to support it.

**Comment by zednis** [comment]
I believe the use case would be to support verification of the authenticity of the download file mentioned in the RDF - much like md5 hashes are used on many open source download pages.

I do think we should consider if that is in an important use case to support.

If this is our use case we need to be able to ensure that the SHA1 (or other) hash is updated every time the associated file is updated.  This may be difficult as we are not the data publisher.

If the hashes are out-of-date then we are not properly supporting the use case as the hashes cannot reliably be used to provide verification of the authenticity of the file.

**Comment by rewolfe** [comment]
USGCRP regularly distributes reports (e.g., the National Climate
Assessment) and images for which a sha1 hash is used to verify the
integrity of the transfer from the archive to the user.  The turtle
(ontology) serialization should support this just like a yaml or json
serialization.

On Wed, Jun 24, 2015 at 4:30 PM, Stephan Zednik notifications@github.com
wrote:

> I believe the use case would be to support verification of the
> authenticity of the download file mentioned in the RDF - much like md5
> hashes are used on many open source download pages.
> 
> I do think we should consider if that is in an important use case to
> support.
> 
> If this is our use case we need to be able to ensure that the SHA1 (or
> other) hash is updated every time the the associated file is updated. This
> may be difficult as we are not the data publisher.
> 
> If the hashes are out-of-date then we are not properly supporting the use
> case as the hashes cannot reliably be used to provide verification of the
> authenticity of the file.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/USGCRP/gcis-ontology/issues/19#issuecomment-115001749
> .

## 

Robert Wolfe, NASA GSFC @ USGCRP, o: 202-419-3470, m: 301-257-6966

**Comment by justgo129** [comment]
Have we reached a consensus on the resolution of this issue?

**Comment by zednis** [comment]
It sounds like we have a use case for supporting it and we will have USGCRP-published resources for which GCIS will be capable of keeping up-to-date SHA1 values.  I am ok with us having hashes for non-USGRCP documents; I think in that case we should investigate having a scheduled process (once per week?) that checks those hashes to discover if our hash value gets out of date - but that is a process concern and not a data model concern.

for representation in the RDF I suggest the following:

```
gcis:cryptographicHash [ 
      gcis:hashFunction dbpedia:SHA-1;
      rdf:value "bdbe9f11d4b77589c541c062cfdeca636440dd82" ] ;
```

It will allow us to support many different hash functions (MD5, SHA-1, SHA-x, etc.) with 2 general properties.

**Comment by justgo129** [comment]
+1

**Comment by rewolfe** [comment]
+1

On Wed, Jul 15, 2015 at 11:19 PM, justgo129 notifications@github.com
wrote:

> +1
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/USGCRP/gcis-ontology/issues/19#issuecomment-121814899
> .

## 

Robert Wolfe, NASA GSFC @ USGCRP, o: 202-419-3470, m: 301-257-6966

**Comment by bduggan** [comment]
On Wednesday, July 15, Stephan Zednik wrote: 

> I think in that case we should investigate having a scheduled process (once per week?) that checks those hashes to discover if our hash value gets out of date - but that is a process concern and not a data model concern.

The gcis-qa tests look at sha1's for nca3 chapters + report:

```
https://github.com/USGCRP/gcis-qa/blob/master/t/06-nca3-manifest.t#L47
```

Currently it runs when a commit is made to gcis-qa (using travis).

Brian

**Comment by justgo129** [comment]
@zednis feel free to merge.

**Comment by justgo129** [comment]
This suggested edits have been addressed at: https://github.com/USGCRP/gcis/pull/201
However, I'm not ready to close this ticket given that we don't have the newly created predicates in the main gcis owl file.  Once those are created, feel free to close.

**Comment by justgo129** [comment]
Closed #19 due to merged https://github.com/USGCRP/gcis/pull/201.
